autopilot/agent.go

```diff
@@ -529,57 +530,135 @@ func (a *Agent) controller() {
 		)
 		pendingMtx.Unlock()
 
-		// If we reach this point, then according to our heuristic we
-		// should modify our channel state to tend towards what it
-		// determines to the optimal state. So we'll call Select to get
-		// a fresh batch of attachment directives, passing in the
-		// amount of funds available for us to use.
-		chanCandidates, err := a.cfg.Heuristic.Select(
-			a.cfg.Self, a.cfg.Graph, availableFunds,
-			numChans, nodesToSkip,
+		// Gather the set of all nodes in the graph, except those we
+		// want to skip.
+		selfPubBytes := a.cfg.Self.SerializeCompressed()
+		var nodes []Node
+		if err := a.cfg.Graph.ForEachNode(func(node Node) error {
+			nID := NodeID(node.PubKey())
+
+			// If we come across ourselves, them we'll continue in
+			// order to avoid attempting to make a channel with
+			// ourselves.
+			if bytes.Equal(nID[:], selfPubBytes) {
+				return nil
+			}
+
+			// Additionally, if this node is in the blacklist, then
+			// we'll skip it.
+			if _, ok := nodesToSkip[nID]; ok {
+				return nil
+			}
+
+			nodes = append(nodes, node)
+			return nil
+		}); err != nil {
+			log.Error("Unable to get graph nodes: %v", err)
+			continue
+		}
+
+		// Use the heuristic to calculate a score for each node in the
+		// graph.
+		scores, err := a.cfg.Heuristic.NodeScores(
+			a.cfg.Graph, nil, availableFunds, nodes,
 		)
 		if err != nil {
-			log.Errorf("Unable to select candidates for "+
-				"attachment: %v", err)
+			log.Errorf("Unable to calculate node scores : %v", err)
 			continue
 		}
 
-		if len(chanCandidates) == 0 {
-			log.Infof("No eligible candidates to connect to")
-			continue
+		// Filter out nodes below our score threshold.
+		numNodes := len(scores)
+		for k, v := range scores {
+			if v.Score < a.cfg.Constraints.MinScoreThreshold {
+				delete(scores, k)
+			}
 		}
 
-		log.Infof("Attempting to execute channel attachment "+
-			"directives: %v", spew.Sdump(chanCandidates))
+		log.Debugf("Got scores for %d nodes (%d below threshold), "+
+			"numChans %d", numNodes, numNodes-len(scores), numChans)
+
+		// Keep opening channels to candidates as long as there are
+		// channel slots open and as long as we have funds available.
+		remChans := numChans
+		remFunds := availableFunds
+		for remChans > 0 && remFunds > 0 {
+			log.Debugf("%d chans remaining, %v funds remaining",
+				remChans, remFunds)
+
+			// To avoid looping forever in case of an error, we
+			// decrease the number of remaining channels now.
+			remChans--
+
+			// Now use the score to make a weighted choice which
+			// node to open a channel to.
+			candidate, err := weightedChoice(scores)
```

Comment:
The proposed design was to select the node (aggregating all the heuristics) that results in the highest "gain". Atm, this will randomly select a node from the single heuristic. is the idea to leave the generic selection algorithm to a follow up PR?

Reply:
Idea is to either here combine scores from different heuristics before making the choice, or create a combined heuristic that abstract this away.